forte/data/data_pack.py

```diff
@@ -478,69 +488,101 @@ def get_span_audio(self, begin: int, end: int) -> np.ndarray:
+                " more AudioPayload in this DataPack"
             )
-        return self._audio[begin:end]
+        return audio_payload_entry.cache[begin:end]
 
     def get_image_array(self, image_payload_idx: int):
-        if image_payload_idx >= len(self.payloads):
-            raise ValueError(
-                f"The input image payload index{(image_payload_idx)}"
-                f" out of range. It should be less than {len(self.payloads)}"
+        """
+        Get the image data in the data pack's payload with specificied image
+        payload index.
+
+        Args:
+            image_payload_idx: the index of the image payload.
+
+        Returns:
+            a numpy array representing the image.
+        """
+        try:
+            image_arr = self.get_single(ImagePayload, image_payload_idx).cache
+        except EntryNotFoundError:
+            raise ProcessExecutionException(
+                "The image payload of this DataPack at index"
+                f"({image_payload_idx}) is not set. Please add"
+                " more ImagePayload in this DataPack"
             )
-        return self.payloads[image_payload_idx]
+        return image_arr
 
     def set_text(
         self,
         text: str,
         replace_func: Optional[Callable[[str], ReplaceOperationsType]] = None,
+        text_payload_index: int = 0,
     ):
+        """
+        Set text for TextPayload at a specified index.
 
-        if len(text) < len(self._text):
-            raise ProcessExecutionException(
-                "The new text is overwriting the original one with shorter "
-                "length, which might cause unexpected behavior."
-            )
-
-        if len(self._text):
-            logging.warning(
-                "Need to be cautious when changing the text of a "
-                "data pack, existing entries may get affected. "
-            )
-
+        Args:
+            text: a str text.
+            replace_func: function that replace text. Defaults to None.
+            text_payload_index (int, optional): TextPayload index in the
+                DataPack. Defaults to 0.
+        """
         span_ops = [] if replace_func is None else replace_func(text)
 
         # The spans should be mutually exclusive
         (
-            self._text,
-            self.__replace_back_operations,
-            self.__processed_original_spans,
-            self.__orig_text_len,
+            text,
+            replace_back_operations,
+            processed_original_spans,
+            orig_text_len,
         ) = data_utils_io.modify_text_and_track_ops(text, span_ops)
 
-    def set_audio(self, audio: np.ndarray, sample_rate: int):
+        tp = TextPayload(self, text_payload_index)
+
+        tp.set_cache(text)
+        tp.meta = TextReadingMeta(self)
+
+        tp.meta.replace_back_operations = replace_back_operations
+        tp.meta.processed_original_spans = processed_original_spans
+        tp.meta.orig_text_len = orig_text_len
```

Comment:
I feel like the `meta` should not be used to store everything we have. @mylibrar could you consider what to do with this?

Reply:
If the replace operations are specific to one `TextPayload`, then I guess maybe we should bound them to that payload?